File: c/quick_sort.c
#include <stdio.h>

static void swap(int *a, int *b) {
    int tmp = *a;
    *a = *b;
    *b = tmp;
}

static void quick_sort(int *items, int start, int end) {
    if (end - start < 2) {
        return;
    }

    int medianIndex = start;

    for (int i = start + 1; i < end; i++) {
        if (items[i] <= items[medianIndex]) {
            if (i > medianIndex + 1) {
                swap(&items[medianIndex], &items[medianIndex + 1]);
            }

            swap(&items[i], &items[medianIndex]);
            medianIndex++;
        }
    }

    quick_sort(items, start, medianIndex);
    quick_sort(items, medianIndex + 1, end);
}

void main() {
    int items[] = { 10, 9, 3, 3, 4, 4, 3, 8, 7, 6, 5, 4, 3, 2, 1, 0, -1 };

    quick_sort(items, 0, 17);

    for (int i = 0; i < 17; i++) {
        printf("%d ", items[i]);
    }
    printf("\n");
}

File: go/quick_sort.go
package main

func quickSort(items []int) {
	if len(items) <= 1 {
		return
	}

	medianIndex := len(items) / 2

	items[0], items[medianIndex] = items[medianIndex], items[0]
	medianIndex = 0

	for i := 1; i < len(items); i++ {
		if items[i] < items[medianIndex] {
			items[i], items[medianIndex] = items[medianIndex], items[i]
			medianIndex++
		}
	}

	quickSort(items[:medianIndex])
	quickSort(items[medianIndex+1:])
}
